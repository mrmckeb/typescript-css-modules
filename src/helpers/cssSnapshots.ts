import { extractICSS, IICSSExports } from 'icss-utils';
import * as postcss from 'postcss';
import * as postcssIcssSelectors from 'postcss-icss-selectors';
import * as ts_module from 'typescript/lib/tsserverlibrary';
import * as less from 'less';
import * as sass from 'sass';
import * as reserved from 'reserved-words';
import { transformClasses } from './classTransforms';
import { Options } from '../options';

const NOT_CAMELCASE_REGEXP = /[\-_]/;
const processor = postcss(postcssIcssSelectors({ mode: 'local' }));

const classNameToProperty = (className: string) => `'${className}': string;`;
const classNameToNamedExport = (className: string) =>
  `export const ${className}: string;`;

const flattenClassNames = (
  previousValue: string[] = [],
  currentValue: string[],
) => previousValue.concat(currentValue);

export const enum FileTypes {
  scss = 'scss',
  less = 'less',
}

export const getFileType = (fileName: string) =>
  fileName.endsWith('less') ? FileTypes.less : FileTypes.scss;

export const getClasses = (css: string, fileType: FileTypes) => {
  try {
    let transformedCss = '';

    if (fileType === FileTypes.less) {
      less.render(css, { asyncImport: true } as any, (err, output) => {
        transformedCss = output.css.toString();
      });
    } else {
      transformedCss = sass.renderSync({ data: css }).css.toString();
    }

    const processedCss = processor.process(transformedCss);

    return extractICSS(processedCss.root).icssExports;
  } catch (e) {
    return {};
  }
};

export const createExports = (classes: IICSSExports, options: Options) => {
  const isCamelCase = (className: string) =>
    !NOT_CAMELCASE_REGEXP.test(className);
  const isReservedWord = (className: string) => !reserved.check(className);

  const processedClasses = Object.keys(classes)
    .map(transformClasses(options.camelCase))
    .reduce(flattenClassNames, []);
  const camelCasedKeys = processedClasses
    .filter(isCamelCase)
    .filter(isReservedWord)
    .map(classNameToNamedExport);

  const defaultExport = `\
declare const classes: {
  ${processedClasses.map(classNameToProperty).join('\n  ')}
};
export default classes;
`;

  if (camelCasedKeys.length) {
    return defaultExport + camelCasedKeys.join('\n') + '\n';
  }
  return defaultExport;
};

export const getDtsSnapshot = (
  ts: typeof ts_module,
  fileName: string,
  scriptSnapshot: ts.IScriptSnapshot,
  options: Options,
) => {
  const css = scriptSnapshot.getText(0, scriptSnapshot.getLength());
  const fileType = getFileType(fileName);
  const classes = getClasses(css, fileType);
  const dts = createExports(classes, options);
  return ts.ScriptSnapshot.fromString(dts);
};
