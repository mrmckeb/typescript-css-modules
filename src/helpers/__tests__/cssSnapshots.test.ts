import { readFileSync } from 'fs';
import { IICSSExports } from 'icss-utils';
import { join } from 'path';
import * as postcss from 'postcss';
import * as postcssIcssSelectors from 'postcss-icss-selectors';
import { createExports, getClasses } from '../cssSnapshots';

const testFileNames = [
  'test.module.css',
  'test.module.less',
  'test.module.scss',
  'empty.module.less',
  'empty.module.scss',
];

const processor = postcss([postcssIcssSelectors({ mode: 'local' })]);

describe('utils / cssSnapshots', () => {
  testFileNames.forEach((fileName) => {
    let classes: IICSSExports;
    const fullFileName = join(__dirname, 'fixtures', fileName);
    const testFile = readFileSync(fullFileName, 'utf8');

    beforeAll(() => {
      classes = getClasses(processor, testFile, fullFileName);
    });

    describe(`with file '${fileName}'`, () => {
      describe('getClasses', () => {
        it('should return an object matching expected CSS', () => {
          expect(classes).toMatchSnapshot();
        });
      });

      describe('createExports', () => {
        it('should create an exports file', () => {
          const exports = createExports(classes, {});
          expect(exports).toMatchSnapshot();
        });
      });
    });
  });
});
