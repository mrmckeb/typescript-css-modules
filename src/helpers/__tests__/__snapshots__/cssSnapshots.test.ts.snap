// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils / cssSnapshots with file 'empty.module.less' createExports should create an exports file 1`] = `
"declare const classes: {
  
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'empty.module.less' getClasses should return an object matching expected CSS 1`] = `Object {}`;

exports[`utils / cssSnapshots with file 'empty.module.scss' createExports should create an exports file 1`] = `
"declare const classes: {
  
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'empty.module.scss' getClasses should return an object matching expected CSS 1`] = `Object {}`;

exports[`utils / cssSnapshots with file 'test.module.css' createExports should create an exports file 1`] = `
"declare const classes: {
  'classA': string;
  'ClassB': string;
  'class-c': string;
  'parent': string;
  'childA': string;
  'childB': string;
  'nestedChild': string;
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'test.module.css' getClasses should return an object matching expected CSS 1`] = `
Object {
  "ClassB": "file__ClassB---2bPVi",
  "childA": "file__childA---1hjQD",
  "childB": "file__childB---pq4Ks",
  "class-c": "file__class-c---DZ1TD",
  "classA": "file__classA---2xcnJ",
  "nestedChild": "file__nestedChild---2d15b",
  "parent": "file__parent---1ATMj",
}
`;

exports[`utils / cssSnapshots with file 'test.module.less' createExports should create an exports file 1`] = `
"declare const classes: {
  'nested-class-parent': string;
  'child-class': string;
  'selector-blue': string;
  'selector-green': string;
  'selector-red': string;
  'column-1': string;
  'column-2': string;
  'column-3': string;
  'column-4': string;
  'color-set': string;
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'test.module.less' getClasses should return an object matching expected CSS 1`] = `
Object {
  "child-class": "file__child-class---1mwoB",
  "color-set": "file__color-set---9sHH_",
  "column-1": "file__column-1---vHRb_",
  "column-2": "file__column-2---28y1r",
  "column-3": "file__column-3---1PsZw",
  "column-4": "file__column-4---2qaaI",
  "nested-class-parent": "file__nested-class-parent---_ft7G",
  "selector-blue": "file__selector-blue---3mslq",
  "selector-green": "file__selector-green---143xX",
  "selector-red": "file__selector-red---Gckob",
}
`;

exports[`utils / cssSnapshots with file 'test.module.scss' createExports should create an exports file 1`] = `
"declare const classes: {
  'local-class-inside-global': string;
  'local-class': string;
  'local-class-2': string;
  'local-class-inside-local': string;
  'nested-class-parent': string;
  'child-class': string;
  'nested-class-parent--extended': string;
  'section-1': string;
  'section-2': string;
  'section-3': string;
  'section-4': string;
  'section-5': string;
  'section-6': string;
  'section-7': string;
  'section-8': string;
  'section-9': string;
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'test.module.scss' getClasses should return an object matching expected CSS 1`] = `
Object {
  "child-class": "file__child-class---1QWYM",
  "local-class": "file__local-class---3SW3k",
  "local-class-2": "file__local-class-2----c5z7",
  "local-class-inside-global": "file__local-class-inside-global---1T0um",
  "local-class-inside-local": "file__local-class-inside-local---1Z9pB",
  "nested-class-parent": "file__nested-class-parent---3qXdF",
  "nested-class-parent--extended": "file__nested-class-parent--extended---qsVau",
  "section-1": "file__section-1---1IHCS",
  "section-2": "file__section-2---cLFhf",
  "section-3": "file__section-3---1ldKa",
  "section-4": "file__section-4---2u0CG",
  "section-5": "file__section-5---1lAYL",
  "section-6": "file__section-6---2YZ9I",
  "section-7": "file__section-7---3w-OF",
  "section-8": "file__section-8---3RB8g",
  "section-9": "file__section-9---3_Mtj",
}
`;
